torch/csrc/jit/passes/peephole_non_tensor.cpp

```diff
@@ -7,6 +7,131 @@
 namespace torch {
 namespace jit {
 
+namespace {
+
+c10::optional<IntAttr::ValueType> getConstantInt(Node& node, size_t pos) {
+  auto n = node.inputs().at(pos)->node();
+  if (n->kind() == prim::Constant &&
+      n->kindOf(attr::value) == AttributeKind::i) {
+    return n->i(attr::value);
+  }
+
+  return {};
+}
+
+c10::optional<IntAttr::ValueType> normalizeArithNode(Node& node) {
+  if (node.inputs().size() != 2) {
+    return {};
+  }
+
+  if (node.kind() == aten::mul || node.kind() == aten::add) {
+    if (auto i = getConstantInt(node, 0)) {
+      node.permuteInputs({1, 0});
+      return i;
+    }
+  }
+
+  return getConstantInt(node, 1);
+}
+
+bool tryReduceArith(Node& node);
+
+/**
+ * Remove a mul/floordiv node if it is multiplication or division by 1.
+ *
+ * @pre node is either aten::mul or aten::floordiv
+ * @return true only if the input IR node is successfully replaced.
+ */
+bool tryReduceMulAndDiv(Node& node) {
+  auto constant = normalizeArithNode(node);
+  if (!constant || *constant != 1) {
+    return false;
+  }
+
+  const auto uses = node.output()->uses();
+  for (const auto& u : uses) {
+    if (tryReduceArith(*u.user)) {
```

Comment:
Im a little confused by the flow here. We iterate through every node in the graph below in `PeepholeOptimizeNonTensorImpl::optimizeBlock`. That invokes, `tryReduceArith` which makes sense. Now, when we find the node, we also look forward into its future uses which we will be calling in `optimizeBlock` anyway. Is there a reason for this ?